### PR TITLE
test: add deterministic tests for the built-in bridge

### PR DESCRIPTION
**Pull request by jarhodes314**
## Proposed changes
This is an attempt at testing the built-in bridge without relying on a real MQTT broker and client. This has a few advantages:
- It's faster than spinning up an MQTT broker
- It's not performance sensitive. I don't know fully how rumqttc tries to deal with excessive load, but I believe our existing tests sometimes put it under more pressure than its happy with, and this causes the connections to fail.
- It's deterministic - if we need to reproduce errors, we can do this directly, we don't have to merely hope that we reproduce the edge case we are attempting to test

Some of the logic we want to test concerns connecting to the broker, this is not something we can cover here. However, this approach should allow us to perform simpler integration tests that are much less likely to suffer from flakiness than the existing tests.

It doesn't yet use proptest, but I would like to try and do this, although probably that will be in a separate PR following this.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 1]
```diff
@@ -442,8 +449,8 @@ impl BridgeMessageSender {
 /// is dropped.
 #[allow(clippy::too_many_arguments)]
 async fn half_bridge(
-    mut recv_event_loop: EventLoop,
-    recv_client: AsyncClient,
+    mut recv_event_loop: Box<dyn MqttEvents>,
+    recv_client: Box<dyn MqttClient>,
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs]: Yes, I don't quite know why I chose to do it this way. I've now swapped this for generics, which removes the need for `clone_dyn`

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 2]
```diff
@@ -591,10 +640,10 @@ fn overall_status(lhs: Option<Status>, rhs: &Option<Status>) -> Option<Status> {
 }
 
 /// A tool to remove duplicate messages and avoid infinite loops
-struct MessageLoopBreaker<Ack, Clock> {
+struct MessageLoopBreaker<'a, Ack: ?Sized, Clock> {
     forwarded_messages: VecDeque<(Instant, Publish)>,
     bidirectional_topics: Vec<Cow<'static, str>>,
-    client: Ack,
+    client: &'a Ack,
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs]: This is another `dyn` issue, so this weirdness has gone away. The logic behind this change originally is `dyn Trait: Trait`, but `Box<dyn Trait>` does not, thus we need to cast `Box<dyn Trait>` to `&dyn Trait` and accept `&Trait` in the function.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 3]
```diff
@@ -944,4 +986,412 @@ mod tests {
...
+            let local_events = [inc!(publish(first_msg)), inc!(publish(second_msg))];
+            let cloud_events = [
+                out!(publish(1)),
+                inc!(puback(1)),
+                // Simulate the cloud sending a second acknowledgement;s
+                inc!(puback(1)),
+                // Then publish the second message
+                out!(publish(2)),
+                inc!(puback(2)),
+            ];
```
Comment (didier-wenzek) [inline crates/extensions/tedge_mqtt_bridge/src/lib.rs:1174]: Looking at this example (duplicated ack) and the two previous ones (inverted ack and disconnection before ack),
I think this is where generated sequences of events would help a lot.

- Publishing 2 local messages
- Expecting 2 local acks 
- For any random sequence of  `out!(publish)`, `inc!(puback)` and `inc!(network_error)` cloud side.
- Not really random sequences as there is some logic relating these messages ... but we should be able to implement filter out sequences making no sense.

**Review by didier-wenzek — APPROVED** [review]
A very nice set of tests and testing ideas, nicely implemented.

Approved, although with some questions.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 4]
```diff
@@ -944,4 +986,412 @@ mod tests {
...
+                bridge.local_client.next_action().unwrap(),
+                Action::Ack(incoming_msg)
+            )
+        }
+
+        #[tokio::test]
+        async fn subscribe_does_not_block() {
+            let (tx, rx) = oneshot::channel();
+            let local_events = UnblockingEventStream::new([inc!(connack)], tx);
+            let client = BlockingSubscribeClient::new(rx);
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs:1078]: Your understanding is correct. I'm not convinced this is white-box, however. Here, we are testing our implementation against the rumqttc interface, which we know requires us to call subscribe when we reconnect and requires the event loop to be polled frequently, and this is exactly what we are testing here. That said, it probably is useful for another test I am currently attempting to write to simulate more accurately the behaviour of rumqttc, where the client the corresponding event to the event loop via a (bounded) channel.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 5]
```diff
@@ -944,4 +986,412 @@ mod tests {
...
+            let cloud_events = [
+                out!(publish(1)),
+                // Abruptly disconnect client
+                inc!(network_error),
+                // Republish message after disconnect
+                out!(publish(1)),
+                inc!(puback(1)),
+                // Then check we successfully acknowledge a future message with the same pkid
+                out!(publish(1)),
+                inc!(puback(1)),
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs:1141]: > I think we would observe this in practice if we had a maximum of 1 in-flight message, rather than the 100 (I believe) we configure rumqttc for.

True. I remember now that indeed rumqttc assigns pkid from 1 to max-in-flight.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs:1078]: >  I am currently attempting to write to simulate more accurately the behaviour of rumqttc

This can be done later. This PR is already a nice improvement.

File: crates/extensions/tedge_mqtt_bridge/src/backoff.rs [hunk 6]
```diff
@@ -78,7 +78,7 @@ mod tests {
         let now = Instant::now();
         let clock = IterClock::new([now, now + Duration::from_secs(1)]);
         let mut backoff = deterministic_backoff(&clock);
-        assert_eq!(backoff.backoff(), Duration::from_secs(30));
+        assert_eq!(backoff.backoff().as_secs(), 30);
```
Comment (jarhodes314) [inline crates/extensions/tedge_mqtt_bridge/src/backoff.rs:81]: I've updated these tests following failures like this when running `cargo stress`:

```
failures:

---- backoff::tests::backoff_is_30_seconds_after_exceeding_timeout stdout ----
thread 'backoff::tests::backoff_is_30_seconds_after_exceeding_timeout' panicked at crates/extensions/tedge_mqtt_bridge/src/backoff.rs:124:9:
assertion `left == right` failed
  left: 30.000000001s
 right: 30s
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: crates/core/tedge_actors/src/runtime.rs [hunk 7]
```diff
@@ -589,7 +589,7 @@ mod tests {
             .expect("Actor to panic in time");
         assert_eq!(
             error.map(|s| s.replace(char::is_numeric, "")), // ignore the task id
-            Some("task  panicked".to_string())
+            Some("task  panicked with message \"Oh dear\"".to_string())
```
Comment (jarhodes314) [inline crates/core/tedge_actors/src/runtime.rs:592]: This is caused by the tokio update

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs:1074]: Didier's comment really helped in understanding what this test is trying to validate. It'd be nice if this commentary is added to the test as well for future readers.

That being said, I was wondering if the test could be made a bit more straight-forward as follows(since the intent is only to validate that the event loop is not blocked by a subscription attempt):
* Include an additional local publish event in the local event stream, followed by the `connack` event. 
* Assert that this local event is still processed by the bridge(asserting the corresponding cloud publish event), while the `subscribe_many` call is is still blocked

Wouldn't that be sufficient? Or I'm missing some other subtle interactions?

File: crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+}
+
+#[derive(Clone)]
+/// Designed to be used with [BlockingSubscribeClient]
+///
+/// Polling this will unblock the call to [BlockingSubscribeClient::subscribe_many]
+pub struct UnblockingEventStream {
+    pub events: FixedEventStream,
+    pub tx: Arc<TokioMutex<Option<oneshot::Sender<EventsPolled>>>>,
+}
```
Comment (albinsuresh) [inline crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs:143]: Since these structs are used exclusively for one single test, wouldn't it be better to keep these structs along with that test? Unless you feel that these can still be used for more tests. 

Initially I had a similar impression of the `ChannelClient` and `ChannelEvents` duo as well, but I see more scope for reuse for them compared to these.

File: crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+#[async_trait::async_trait]
+impl AllProcessed for UnblockingEventStream {
+    async fn all_processed(&self) -> anyhow::Result<()> {
+        self.events.all_processed().await
+    }
+}
+
+#[derive(Clone)]
+/// A client that panics on every operation
+pub struct PanickingClient;
```
Comment (albinsuresh) [inline crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs:163]: I couldn't find any usages of this panicking client in any tests. Did you accidentally remove any tests that were using this?

**Review by albinsuresh — COMMENTED** [review]
Nicely done. Most of the tests were really straight-forward with the newly added abstractions.

File: crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+macro_rules! inc {
+    (connack) => {
+        Ok(Event::Incoming(Incoming::ConnAck(ConnAck {
+            session_present: true,
+            code: ConnectReturnCode::Success,
+        })))
+    };
+    (suback) => {
+        Ok(Event::Incoming(Incoming::SubAck(SubAck {
+            pkid: 1,
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs:41]: It doesn't really matter, we aren't going to interact with it, and I think behaving technically incorrectly is preferable to polluting test code with irrelevant detail.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 11]
```diff
@@ -944,4 +981,487 @@ mod tests {
...
+        async fn health_success_is_not_published_before_client_connected() {
+            let mut bridge = Bridge::default().process_all_events().await;
+            // There should not be a health message since there are no events
+            assert_eq!(bridge.next_health_message(), None);
+        }
+
+        #[tokio::test]
+        async fn health_success_is_published_on_connack() {
+            let mut bridge = Bridge::default()
+                .with_local_events([inc!(connack), inc!(suback)])
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/lib.rs:1200]: It does have some relevance, but that's really testing a separate thing (that we don't publish consecutive "healthy" messages, we only publish one when the status changes).

File: crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs (hunk 9 above)
Comment (jarhodes314) [reply crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs]: I believe so, now deleted.

File: crates/extensions/tedge_mqtt_bridge/src/lib.rs [hunk 12]
```diff
@@ -944,4 +981,491 @@ mod tests {
...
+            bridge.assert_not_panicked().await;
+            let expected_message = Publish {
+                topic: "s/us".into(),
+                ..publish
+            };
+            assert_eq!(
+                bridge.cloud_client.next_action().unwrap(),
+                Action::Publish(expected_message)
+            )
+        }
```
Comment (jarhodes314) [inline crates/extensions/tedge_mqtt_bridge/src/lib.rs:1099]: Response to @albinsuresh's comments

> Didier's comment really helped in understanding what this test is trying to validate. It'd be nice if this commentary is added to the test as well for future readers.

I've now added something along those lines as a comment at the top of the test.

> That being said, I was wondering if the test could be made a bit more straight-forward as follows(since the intent is only to validate that the event loop is not blocked by a subscription attempt):
>
> Include an additional local publish event in the local event stream, followed by the connack event.
Assert that this local event is still processed by the bridge(asserting the corresponding cloud publish event), while the subscribe_many call is is still blocked
Wouldn't that be sufficient? Or I'm missing some other subtle interactions?

Also done. This means the `BlockingSubscribeClient` merely needs to block, it doesn't need to communicate with an event loop, since we don't care about the subscription call completing. We can trust that `subscribe_many` is called on `connack` due to the unit test covering that action. The only downside of this approach is that the test could plausibly be accidentally modified to not trigger connack and it would still pass.

**Review by albinsuresh — APPROVED** [review]
LGTM.

**Review by didier-wenzek — APPROVED** [review]
I confirm my approval after the latest improvements.
